src/DS/sds.c

```diff
@@ -246,108 +270,179 @@ static int ds_sds_dump_component(const char* external_file, const char* componen
+
+	return ds_sds_register_component(session, doc, inner_root, component_id, target_filename_dirname, relative_filepath);
+}
+
+static int ds_sds_dump_file_component(const char* external_file, const char* component_id, struct ds_sds_session *session, const char *target_filename_dirname, const char *relative_filepath)
+{
+	int ret = 0;
+
+	struct oscap_source *source_file = load_referenced_source(session, external_file);
+	xmlDoc *doc = oscap_source_get_xmlDoc(source_file);
+
+	if (doc == NULL) {
 		ret = -1;
 		goto cleanup;
 	}
 
-	// If the inner root is script, we have to treat it in a special way
-	if (strcmp((const char*)inner_root->name, "script") == 0) {
-		// the cast is safe to do because we are using the GNU basename, it doesn't
-		// modify the string
-		const char* file_basename = basename((char*)relative_filepath);
-		const char* sce_filename = oscap_sprintf("%s/%s/%s",ds_sds_session_get_target_dir(session), target_filename_dirname, file_basename);
-		ret = ds_sds_dump_component_sce(inner_root->children, component_id, sce_filename);
-		oscap_free(sce_filename);
-		if (ret != 0) {
-			goto cleanup;
-		}
-	}
-	// Otherwise we create a new XML doc we will dump the contents to.
-	// We can't just dump node "innerXML" because namespaces have to be
-	// handled.
-	else {
-		xmlDoc *new_doc = ds_doc_from_foreign_node(inner_root, doc);
-		if (new_doc == NULL) {
-			ret = -1;
-			goto cleanup;
-		}
+	xmlNodePtr inner_root = ds_sds_get_component_root_by_id(doc, component_id);
 
-		struct oscap_source *component_source = oscap_source_new_from_xmlDoc(new_doc, relative_filepath);
-		ds_sds_session_register_component_source(session, relative_filepath, component_source);
+	if (ds_sds_register_component(session, doc, inner_root, component_id, target_filename_dirname, relative_filepath) != 0) {
+		ret = -1;
+		goto cleanup;
 	}
 
 	cleanup:
 		oscap_source_free(source_file);
 		return ret;
 }
 
-int ds_sds_dump_component_ref_as(const xmlNodePtr component_ref, struct ds_sds_session *session, const char* sub_dir, const char* relative_filepath)
+static int ds_dsd_dump_remote_component(const char* url, const char* component_id, struct ds_sds_session *session, const char *target_filename_dirname, const char *relative_filepath)
 {
-	char* cref_id = (char*)xmlGetProp(component_ref, BAD_CAST "id");
-	if (!cref_id)
-	{
-		oscap_seterr(OSCAP_EFAMILY_XML, "No or invalid id attribute on given component-ref.");
-		xmlFree(cref_id);
+	int ret = 0;
+	size_t memory_size = 0;
+
+	ds_sds_session_remote_resources_progress(session)(false, "Downloading: %s ... ", url);
+
+	char* mem = oscap_acquire_url_download(url, &memory_size);
+	if (mem == NULL) {
+		ds_sds_session_remote_resources_progress(session)(false, "error", url);
 		return -1;
 	}
 
-	char* xlink_href = (char*)xmlGetNsProp(component_ref, BAD_CAST "href", BAD_CAST xlink_ns_uri);
+	ds_sds_session_remote_resources_progress(session)(false, "ok", url);
```

Comment:
I think there should be "ok\n", because when I run `oscap info`, I got messed output:
```
Checklists:
	Ref-Id: scap_gov.nist_cref_r2910-xccdf.xml
Downloading: http://scap.nist.gov/validation/downloads/r2910-external-oval-component.xml ... ok		Status: draft
```